image_segmentation/pytorch/runtime/training.py

```diff
@@ -45,82 +45,86 @@ def train(flags, model, train_loader, val_loader, loss_fn, score_fn, device, cal
-            if flags.amp:
-                scaler.scale(loss_value).backward()
-            else:
-                loss_value.backward()
+                with autocast(enabled=flags.amp):
+                    output = model(curr_image)
+                    loss_value = loss_fn(output, curr_label)
+                    loss_value /= flags.ga_steps
 
-            accumulated_steps += 1
-            if accumulated_steps % flags.ga_steps == 0:
                 if flags.amp:
-                    scaler.step(optimizer)
-                    scaler.update()
+                    scaler.scale(loss_value).backward()
                 else:
-                    optimizer.step()
+                    loss_value.backward()
 
-                optimizer.zero_grad()
-                accumulated_steps = 0
+            if flags.amp:
+                scaler.step(optimizer)
+                scaler.update()
+            else:
+                optimizer.step()
+
+            optimizer.zero_grad()
 
             loss_value = reduce_tensor(loss_value, world_size).detach().cpu().numpy()
             cumulative_loss.append(loss_value)
 
-        mllog_end(key=CONSTANTS.EPOCH_STOP, sync=True,
+        mllog_end(key=CONSTANTS.EPOCH_STOP, sync=False,
                   metadata={CONSTANTS.EPOCH_NUM: epoch, 'current_lr': optimizer.param_groups[0]['lr']})
 
         if flags.lr_decay_epochs:
             scheduler.step()
+
         if ((epoch % flags.evaluate_every) == 0) and epoch >= flags.start_eval_at:
             del output
-            mllog_start(key=CONSTANTS.EVAL_START, value=epoch, metadata={CONSTANTS.EPOCH_NUM: epoch}, sync=True)
+            mllog_start(key=CONSTANTS.EVAL_START, value=epoch, metadata={CONSTANTS.EPOCH_NUM: epoch}, sync=False)
 
             eval_metrics = evaluate(flags, model, val_loader, loss_fn, score_fn, device, epoch)
             eval_metrics["train_loss"] = sum(cumulative_loss) / len(cumulative_loss)
 
             mllog_event(key=CONSTANTS.EVAL_ACCURACY,
-                        value={"epoch": epoch, "value": eval_metrics["mean_dice"]},
+                        value=eval_metrics["mean_dice"],
                         metadata={CONSTANTS.EPOCH_NUM: epoch},
                         sync=False)
-            mllog_end(key=CONSTANTS.EVAL_STOP, metadata={CONSTANTS.EPOCH_NUM: epoch}, sync=True)
+            mllog_end(key=CONSTANTS.EVAL_STOP, metadata={CONSTANTS.EPOCH_NUM: epoch}, sync=False)
 
             for callback in callbacks:
-                callback.on_epoch_end(epoch, eval_metrics, model, optimizer)
+                callback.on_epoch_end(epoch=epoch, metrics=eval_metrics, model=model, optimizer=optimizer)
             model.train()
             if eval_metrics["mean_dice"] >= flags.quality_threshold:
                 stop_training = True
 
-        mllog_end(key=CONSTANTS.BLOCK_STOP, sync=True,
+        mllog_end(key=CONSTANTS.BLOCK_STOP, sync=False,
                   metadata={CONSTANTS.FIRST_EPOCH_NUM: epoch, CONSTANTS.EPOCH_COUNT: 1})
```

Comment:
should this be the size of the block,
CONSTANTS.BLOCK_START - CONSTANTS.BLOCK_STOP ? e.g. 1313 for the first eval block.

Same thing for the epoch count at CONSTANTS.BLOCK_START

Reply:
I'm not sure I understand. A block is an aggregation of n epochs if you suffer from logging every epoch. 
Could you please provide a code how would you imagine this to look like?

Reply:
Sorry, please ignore my comment.
I see that in the reference BLOCK_START and BLOCK_STOP is printed for every epoch, so the epoch_count=1. This will print below for the first epoch: 
-- BLOCK_START=1, EPOCH_COUNT=1
-- BLOCK_STOP=1, EPOCH_COUNT=1


In my case, I will print this once for the first 1000 epochs, as below. 
e.g. 
-- BLOCK_START=0, EPOCH_COUNT=1000
-- BLOCK_STOP=1000, EPOCH_COUNT=1000

Reply:
Yes, that would be accurate